Runtime/C++/src/bebop.hpp

```diff
@@ -0,0 +1,292 @@
+#pragma once
+
+#include <chrono>
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+#include <memory>
+#include <string>
+#include <vector>
+
+#define BEBOP_ASSUME_LITTLE_ENDIAN 1
+
+namespace bebop {
+
+/// A "tick" is a ten-millionth of a second, or 100ns.
+using Tick = std::ratio<1, 10000000>;
+using TickDuration = std::chrono::duration<int64_t, Tick>;
+
+namespace {
+    /// The number of ticks between 1/1/0001 and 1/1/1970.
+    const int64_t ticksBetweenEpochs = 621355968000000000;
+}
+
+enum class GuidStyle {
+    Dashes,
+    NoDashes,
+};
+
+struct Guid {
+    /// The GUID bytes, in .NET "ToByteArray" order.
+    /// https://docs.microsoft.com/en-us/dotnet/api/system.guid.tobytearray?view=net-5.0
+    uint8_t m_bytes[16];
+
+    Guid() = default;
+    Guid(const uint8_t *bytes) { std::memcpy(m_bytes, bytes, 16); }
+    Guid(Guid const &other) : Guid(other.m_bytes) {}
+
+    static Guid fromString(const std::string &string) {
+        Guid g;
+        static const uint8_t asciiToHex[256] = {
+            0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+            0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+            0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+            0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  0,  0,  0,  0,  0,  0,
+            0, 10, 11, 12, 13, 14, 15,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+            0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+            0, 10, 11, 12, 13, 14, 15,  // and the rest is zeroes
+        };
+        const char* s = string.c_str(); uint8_t a;
+
+#define Nibble() (asciiToHex[static_cast<unsigned char>(*s++)])
```

Comment:
I wonder if you could do something like:
```C++
static auto Nibble = [](const uint8_t* asciiToHex, const char*& s) { asciiToHex[static_cast<unsigned char>(*s++)]; };
```
Then you could avoid having to #undef Nibble and potentially avoid some other header file that happens to be included with this header (in some future application) #defining Nibble and causing a confusing compiler error.

Reply:
Same comment for Dash and B.

Reply:
Not sure if my lambda example function for Nibble exactly captures what your macro is doing, but the point being that it would be safer to use a lambda instead of a macro.